include/deal.II/base/exceptions.h

```diff
@@ -1762,6 +1762,33 @@ namespace deal_II_exceptions
 #endif /*ifdef DEAL_II_HAVE_BUILTIN_EXPECT*/
 
 
+/**
+ * `DEAL_II_NOT_IMPLEMENTED` is a macro (that looks like a function call
+ * when used as in `DEAL_II_NOT_IMPLEMENTED();`) that is used to raise an
+ * error in places where a piece of code is not yet implemented. If a code
+ * runs into such a place, it will be aborted with an error message that
+ * explains the situation, along with a backtrace of how the code ended
+ * up in this place.
+ *
+ * Alternatively, if
+ * deal_II_exceptions::internals::ExceptionHandling::abort_or_throw_on_exception
+ * is set to ExceptionHandling::throw_on_exception, then the corresponding
+ * error is thrown as a C++ exception that can be caught (though in
+ * many cases codes will then find it difficult to do what they wanted
+ * to do).
+ */
+#define DEAL_II_NOT_IMPLEMENTED()                                \
+  ::dealii::deal_II_exceptions::internals::issue_error_noreturn( \
+    ::dealii::deal_II_exceptions::internals::ExceptionHandling:: \
+      abort_or_throw_on_exception,                               \
+    __FILE__,                                                    \
+    __LINE__,                                                    \
+    __PRETTY_FUNCTION__,                                         \
+    nullptr,                                                     \
+    nullptr,                                                     \
+    ExcNotImplemented())
+
```

Comment:
Does this work on GPUs? We should probably call `Kokkos::abort(error_message);` if we are on a device similar to what we are doing `Assert` already.

Reply:
I can make that change -- would you be willing to let me do this in a follow-up?

Out of curiosity, what happens if you call `std::abort()` on a GPU?

Reply:
> Out of curiosity, what happens if you call std::abort() on a GPU?

It wouldn't compile, see https://godbolt.org/z/T87Pes3ox.

Reply:
> I can make that change -- would you be willing to let me do this in a follow-up?

Sure, it seems we don't have any problematic use cases in our CI yet.